controller.go

```diff
@@ -185,19 +189,54 @@ func (c *controller) validateHostDiscoveryConfig() bool {
 	return true
 }
 
-func (c *controller) initDiscovery() error {
+func (c *controller) initDiscovery(watcher discovery.Watcher) error {
 	if c.cfg == nil {
 		return fmt.Errorf("discovery initialization requires a valid configuration")
 	}
 
-	hostDiscovery := hostdiscovery.NewHostDiscovery()
-	return hostDiscovery.StartDiscovery(&c.cfg.Cluster, c.hostJoinCallback, c.hostLeaveCallback)
+	c.discovery = hostdiscovery.NewHostDiscovery(watcher)
+	return c.discovery.Watch(c.hostJoinCallback, c.hostLeaveCallback)
 }
 
-func (c *controller) hostJoinCallback(hosts []net.IP) {
+func (c *controller) hostJoinCallback(nodes []net.IP) {
+	c.processNodeDiscovery(nodes, true)
 }
 
-func (c *controller) hostLeaveCallback(hosts []net.IP) {
+func (c *controller) hostLeaveCallback(nodes []net.IP) {
+	c.processNodeDiscovery(nodes, false)
+}
+
+func (c *controller) processNodeDiscovery(nodes []net.IP, add bool) {
+	var self net.IP
+	if c.cfg != nil {
+		addr := strings.Split(c.cfg.Cluster.Address, ":")
+		self = net.ParseIP(addr[0])
+	}
+	c.Lock()
+	drivers := c.drivers
+	c.Unlock()
+
+	for _, d := range drivers {
+		if d.capability.DataScope != datastore.GlobalScope {
+			continue
+		}
+		for _, node := range nodes {
+			nodeData := driverapi.NodeDiscoveryData{Address: node.String(), Self: node.Equal(self)}
+			jsonStr, err := json.Marshal(nodeData)
+			if err != nil {
+				log.Errorf("%v", err)
```

Comment:
Can we add a bit of context on this error message? Otherwise it will look like some low level marshalling error and nobody will know where it is coming from